### PR TITLE
[MPS] Build only the reduced Q in linalg.qr (fix tall-skinny perf cliff)

### LINKED ISSUE #187567: [MPS] torch.linalg.qr is 25-250× slower than CPU on tall-skinny matrices
## Description

`torch.linalg.qr` on MPS scales super-linearly with the row dimension for tall-skinny inputs `(n, m)` where `n >> m`. CPU `torch.linalg.qr` on the same input scales roughly linearly. The slowdown widens with `n` — at `n=4000, m=36` the MPS kernel is **253× slower** than the CPU LAPACK call. Even a `MPS → CPU → QR → MPS` round-trip is **47-170× faster than the native MPS path** at every size measured.

Symptoms suggest the MPS kernel computes the **full** QR (Q is `n × n`) rather than the **reduced** QR (Q is `n × m`), since the empirical scaling fits roughly `O(n^1.6)` — close to what an `O(n × n × m)` Householder accumulation would predict, but very far from the `O(n × m²)` reduced QR.

This silently makes any algorithm that orthonormalises tall-skinny blocks on MPS — block Krylov solvers, LOBPCG, randomized SVD, block conjugate-gradient — drastically slower than the equivalent on CPU.

## Minimum reproduction

```python
import time
import torch

n_cols = 36
sizes = [200, 500, 1000, 1500, 2000, 3000, 4000]

def bench(label, fn, n, reps=10):
    torch.manual_seed(0)
    Z_template = torch.randn(n, n_cols, dtype=torch.float32)
    fn(Z_template.clone().to("mps") if "MPS" in label else Z_template.clone())  # warm up
    times = []
    for _ in range(reps):
        Z = Z_template.clone()
        if "MPS" in label or "round" in label:
            Z = Z.to("mps")
        torch.mps.synchronize()
        t0 = time.perf_counter()
        fn(Z)
        torch.mps.synchronize()
        times.append(time.perf_counter() - t0)
    times.sort()
    return times[len(times) // 2] * 1e3

print(f"{'n':>6s} {'native MPS':>12s} {'CPU':>10s} {'round-trip':>12s} {'MPS/CPU':>10s}")
for n in sizes:
    mps = bench("native MPS", lambda Z: torch.linalg.qr(Z), n)
    cpu = bench("CPU",        lambda Z: torch.linalg.qr(Z), n)
    rt  = bench("round-trip", lambda Z: torch.linalg.qr(Z.cpu())[0].to("mps"), n)
    print(f"{n:>6d} {mps:>12.2f} {cpu:>10.2f} {rt:>12.2f} {mps/cpu:>10.1f}x")

```

Output
```bash
PyTorch 2.12.0, Apple M4 Mac mini, macOS 25.3:
     n    native MPS (ms)     CPU (ms)    round-trip (ms)    MPS/CPU
----------------------------------------------------------------------
   200               1.98         0.08               0.43       25.9x
   500               5.14         0.14               0.48       38.0x
  1000              13.51         0.25               0.61       53.5x
  1500              34.56         0.34               0.71      100.3x
  2000              61.70         0.45               0.82      135.9x
  3000             133.00         0.68               1.11      194.9x
  4000             229.83         0.91               1.34      252.7x
```

Expected behavior
torch.linalg.qr on MPS should match or approach the CPU scaling for tall-skinny inputs — at worst it should be flat overhead plus the actual reduced-QR work O(n m²). Today it's roughly O(n^1.6) which is consistent with a full-QR implementation.
Environment
PyTorch: 2.12.0
Python: 3.14
Device: Apple M4 Mac mini, macOS 25.3
Backend: MPS (Metal)
Workaround
For now, route tall-skinny QR through CPU:

This is what we're doing in [sparsexlab/torch-sla](https://github.com/sparsexlab/torch-sla) for the LOBPCG eigensolver (relevant PR: sparsexlab/torch-sla#45). End-to-end LOBPCG wall-clock at n=2000, k=6 dropped from 12.5 s to 0.087 s on M4 MPS after this swap.
Related

This may share a root with the missing aten::_linalg_eigh.eigenvalues on MPS — many linalg ops on MPS appear to be either un-tuned or routed through fall-back paths. Filing this one specifically because the O(n²) scaling makes it a silent perf cliff rather than an explicit NotImplementedError.
`

### Versions

PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A
OS: macOS 26.3 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: Could not collect
Libc version: N/A
Python version: 3.14.5 (main, May 10 2026, 10:21:34) [Clang 21.0.0 (clang-2100.0.123.102)] (64-bit runtime)
Python platform: macOS-26.3-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M4
Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] onnxruntime==1.26.0
[pip3] torch==2.12.0
[conda] Could not collect

torch.backends.mps.is_available(): True
torch.backends.mps.is_built(): True

cc @ezyang @gchanan @kadeng @msaroufim @malfet @aditvenk @jianyuh @nikitaved @mruberry @walterddr @Lezcano @kulinseth @DenisVieriu97 @jhavukainen @Isalia20

### PR DESCRIPTION
**Pull request by GodlyDonuts**
Fixes #187567

### Summary

`torch.linalg.qr` on MPS scales super-linearly for tall-skinny inputs because Q construction materializes full `m x m` Householder matrices even when reduced mode only needs an `m x k` result. At `(4000, 36)`, the issue report measured MPS at roughly 250x slower than CPU.

### Current root cause

Since this PR was opened, #189192 split MPS QR into `geqrf` factorization and `orgqr` Q construction. The refactor removed the old combined QR kernel, but `orgqr` still creates three `m x m` workspaces, explicitly builds each dense reflector, and accumulates them with forward matrix multiplications. Reduced QR therefore retains the quadratic working set and super-linear row scaling.

### Approach

Keep the new `geqrf`/`orgqr` architecture and optimize `orgqr`:

1. Allocate one Q workspace with the same shape as the requested output (`m x k` in reduced mode and `m x m` in complete mode).
2. Initialize it to the corresponding columns of the identity matrix.
3. Apply the compact reflectors from last to first as left multiplications, `Q <- H_i Q`.
4. Copy the completed panel to the output only after all reflectors have been consumed.

Each SIMD group owns one Q column during an update. Reduced mode now uses `m x k` scratch and `O(m * k^2)` Q-construction work. Complete mode retains the full output and its unavoidable cost. Batched and strided inputs and the existing real and complex `orgqr` dtypes remain supported.

### Test plan

Added parametrized coverage for reconstruction, orthonormality, triangular R, and output shapes across reduced and complete modes, tall and wide matrices, and a batched input.

Repository checks run:

```
/Users/sairamen/projects/.venv/bin/spin fixlint aten/src/ATen/native/mps/kernels/LinearAlgebra.h aten/src/ATen/native/mps/kernels/LinearAlgebra.metal aten/src/ATen/native/mps/operations/LinearAlgebra.mm test/test_mps.py
/Users/sairamen/projects/.venv/bin/spin lint aten/src/ATen/native/mps/kernels/LinearAlgebra.h aten/src/ATen/native/mps/kernels/LinearAlgebra.metal aten/src/ATen/native/mps/operations/LinearAlgebra.mm test/test_mps.py
/Users/sairamen/projects/.venv/bin/python -m py_compile test/test_mps.py
git diff --check
```

The post-refactor MPS runtime tests were not run locally because the configured virtual environment does not contain a PyTorch build. The requested `ciflow/mps` workflow should cover:

```
python test/test_mps.py -k test_linalg_qr_tall_skinny -v
python test/test_mps.py -k test_householder_product_race -v
```

### CODE REVIEW

**Comment by GodlyDonuts** [comment]
Hi @malfet, gentle follow-up on this triaged MPS QR performance fix. Could you approve the pending MPS workflow and take an initial look when you have a moment? The PR includes targeted correctness coverage; I am happy to rebase or adjust anything needed.

**Comment by GodlyDonuts** [comment]
Hi @kurtamohler @jhavukainen — could either of you review this MPS QR performance fix when you have a chance? The branch is now rebased over #189192 and ports the optimization into the new `orgqr` path: reduced mode builds only an output-shaped Q panel and applies the `geqrf` reflectors backward. Since you authored and reviewed that refactor, your eyes would be especially helpful. The `ciflow/mps` workflow also still needs maintainer approval. Thanks!
